File: 21/haskell/01.hs
-- stack --resolver lts script
readInts :: String -> [Int]
readInts = map read . lines

pairDepths :: [Int] -> [(Int, Int)]
pairDepths depths = zip (init depths) $ tail depths

countIncreases = length . filter (\ (a,b) -> a < b)

pairDepths3 :: [Int] -> [(Int, Int)]
pairDepths3 depths = zip depths $ drop 3 depths

main = do
    contents <- readFile "../INPUT/01"
    let part1 = show . countIncreases . pairDepths . readInts $ contents
    let part2 = show . countIncreases . pairDepths3 . readInts $ contents
    print $ "part1: " ++ part1
    print $ "part2: " ++ part2

File: 21/lua/01.lua
local file = "../INPUT/01"

local increases = -1
local depth = 0

for line in io.lines(file) do
    line = tonumber(line)
    if depth < line then
        increases = increases + 1
    end
    depth = line
end

print("part1:", increases)

local depths = {}
increases = 0

for line in io.lines(file) do
    line = tonumber(line)
    if #depths < 3 then
        table.insert(depths, line)
    else
        --[[ print(depths)
        for i, v in ipairs(depths) do
            print("depth:", i, v)
        end
        print("line:", line) ]]

        local currdepths = depths[1] + depths[2] + depths[3]
        local nextdepths = depths[2] + depths[3] + line
        --[[ print("currdepths:", currdepths)
        print("nextdepths:", nextdepths) ]]

        if nextdepths > currdepths then increases = increases +1 end

        table.remove(depths, 1)
        table.insert(depths, line)
    end
end

print("part2:", increases)
